clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -6297,6 +6297,7 @@ static void processSYCLKernel(Sema &S, FunctionDecl *FD, MangleContext &MC) {
     const CXXRecordDecl *CRD = (KernelParamTy->isReferenceType()
                                     ? KernelParamTy->getPointeeCXXRecordDecl()
                                     : KernelParamTy->getAsCXXRecordDecl());
+    assert(CRD && "invalid kernel caller");
```

Comment:
On device we throw an error, if there is no kernel object. See https://github.com/intel/llvm/blob/5c3e5385b35d3bbf97403f68be64ef24159697cf/clang/lib/Sema/SemaSYCL.cpp#L3097
Should we do the same thing on host?

Reply:
@Fznamznon Thanks for looking into the patch. 

Here for device, we are calling ConstructOpenCLKernel()
 if (S.LangOpts.SYCLIsDevice) {
    S.ConstructOpenCLKernel(FD, MC);
  } 

I see that we are giving assert for no kernel object on ConstructOpenCLKernel() 
Here void Sema::ConstructOpenCLKernel(FunctionDecl *KernelCallerFunc,
                                 MangleContext &MC) {
  // The first argument to the KernelCallerFunc is the lambda object.
  const CXXRecordDecl *KernelObj = getKernelObjectType(KernelCallerFunc);
  assert(KernelObj && "invalid kernel caller");
  // Do not visit invalid kernel object.
  if (KernelObj->isInvalidDecl())
    return;

I think we should do similar on host like device instead of error here.
We might need to add this on host too :
// Do not visit invalid kernel object.
  if (KernelObj->isInvalidDecl())
    return;

What is your opinion on it?

Reply:
Yes, but If I remember correctly, we throw an error inside `CheckSYCLKernelCall` function, and it is called before `ConstructOpenCLKernel`.

Reply:
I see what you meant to say here. I have updated patch to throw an error for no kernel object on host and added extra RUN check for host on test "kernel-not-functor-host.cpp"

Reply:
Thanks Mariya.